src/BaseUriPlugin.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+namespace Http\Client\Plugin;
+
+use Http\Promise\Promise;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\UriInterface;
+
+class BaseUriPlugin implements Plugin
+{
+    /**
+     * @var UriInterface
+     */
+    private $baseUri;
+
+    /**
+     * @param UriInterface $baseUri
+     */
+    public function __construct(UriInterface $baseUri)
+    {
+        $this->baseUri = $baseUri;
+    }
+
+    /**
+     * Handle the request and return the response coming from the next callable.
+     *
+     * @param RequestInterface $request
+     * @param callable         $next    Next middleware in the chain, the request is passed as the first argument
+     * @param callable         $first   First middleware in the chain, used to to restart a request
+     *
+     * @return Promise
+     */
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
+    {
+        $uriString = $request->getUri()->__toString();
+        if (!strpos($uriString, '://')) {
```

Comment:
i would expect a base uri plugin to only trigger if my uri has no domain in it, and never replace an explicitly specified domain. agree with tobias on this one. if there is a real use case to replace domains, we can do a plugin, but it seems like a rather weird edge case where we might better let people write a custom plugin to handle their case. as in replace only some domains or whatever.

@Nyholm another question on this line: what about an uri like `://example.com/foo`, so one that is missing just the protocol part? should we handle that? i tend to say no, but we might want to be explicit in the documentation on that.